src/backend/lib/upgrade-storage.js

```diff
@@ -65,36 +97,50 @@ class Storage {
   // TODO: factor out the 'version' parameter and use 'this.version' instead,
   // since we already have it.
   setApkInfo(apkPath, version, cb) {
-    log("setApkInfo", apkPath, version);
-    this.apkToUpgradeOption(apkPath, version, (err, info) => {
-      if (err) return cb(err);
-      this.currentApk = info;
-      cb();
+    this.lock.writeLock(release => {
+      function done(err) {
+        release();
+        cb(err);
+      }
+      log("setApkInfo", apkPath, version);
+      this.apkToUpgradeOption(apkPath, version, (err, info) => {
+        if (err) return done(err);
+        this.currentApk = info;
+        done();
+      });
     });
   }
 
   // Callback<[UpgradeOption]> -> Void
   getAvailableUpgrades(cb) {
-    log("getAvailableUpgrades");
-    const results = [];
-    if (this.currentApk) {
-      const currentApk = Object.assign({}, this.currentApk);
-      results.push(currentApk);
-    }
-    this.localUpgrades.get((err, options) => {
-      if (err) return cb(err);
-      options.forEach(option => {
-        const opt = Object.assign({}, option);
-        results.push(opt);
+    this.lock.readLock(release => {
+      function done(err, res) {
+        release();
+        cb(err, res);
+      }
+      log("getAvailableUpgrades");
+      const results = [];
+      if (this.currentApk) {
+        const currentApk = Object.assign({}, this.currentApk);
+        results.push(currentApk);
+      }
+      this.localUpgrades.get((err, options) => {
+        if (err) return done(err);
+        options.forEach(option => {
+          const opt = Object.assign({}, option);
+          results.push(opt);
+        });
+        log("getAvailableUpgrades results:", results);
+        done(null, results);
       });
-      log("getAvailableUpgrades results:", results);
-      cb(null, results);
     });
   }
 
   // String, String, Callback<Void> -> WritableStream
   createApkWriteStream(filename, version, hash, cb) {
+    // Why doesn't accept locks?
```

Comment:
What do you mean by this?

Reply:
When I tried using locks with this function the program came to a hault with tests. Only investigated enough to find out the locks there were causing problems, will dig further for more detail.